jujugui/static/gui/src/app/components/copy-to-clipboard/test-copy-to-clipboard.js

```diff
@@ -0,0 +1,90 @@
+the terms of the GNU Affero General Public License version 3, as published by
+the Free Software Foundation.
+
+This program is distributed in the hope that it will be useful, but WITHOUT
+ANY WARRANTY; without even the implied warranties of MERCHANTABILITY,
+SATISFACTORY QUALITY, or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Affero
+General Public License for more details.
+
+You should have received a copy of the GNU Affero General Public License along
+with this program.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+'use strict';
+
+var juju = {components: {}};  // eslint-disable-line no-unused-vars
+var testUtils = React.addons.TestUtils;
+
+chai.config.includeStack = true;
+chai.config.truncateThreshold = 0;
+
+describe('CopyToClipboard', function() {
+  beforeAll(function(done) {
+    // By loading these files it makes their classes available in the tests.
+    YUI().use('copy-to-clipboard', function() { done(); });
+  });
+
+  it('renders with a default value', function() {
+    var output = jsTestUtils.shallowRender(
+        <juju.components.CopyToClipboard/>);
+    var className = output.props.className;
+    var expected = (
+      <div id={output.props.id} className={className}>
+        <input className={className + '__input'}
+               ref="input"
+               readOnly="true"
+               value=""/>
+        <button className={className + '__btn'}
+                ref="btn">
+          <juju.components.SvgIcon name="get-link-url_16"
+            size="16" />
+        </button>
+      </div>
+    );
+    assert.deepEqual(output, expected, 'Did not render as expected');
+  });
+
+  it('renders a user-provided value properly', function() {
+    var value = 'foobar';
+    var output = testUtils.renderIntoDocument(
+        <juju.components.CopyToClipboard value={value}/>);
+    assert.equal(output.refs.input.value, value,
+                 'Value is not set properly for input');
+  });
+
+  it('initializes the Clipboard widget', function() {
+    var realClipboard = window.Clipboard;
+    window.Clipboard = sinon.spy();
+    var component = testUtils.renderIntoDocument(
+        <juju.components.CopyToClipboard/>);
+    var className = '.' + component.props.className + '__btn';
+    assert.include(Clipboard.getCall(0).args[0], className,
+                   'Clipboard was not initialized with expected selector');
+    window.Clipboard = realClipboard;
+  });
+
+  it('supports multiple components on the same page', function() {
+    var components = [];
+    for (var i = 0; i < 3; i++) {
+      components.push(testUtils.renderIntoDocument(
+          <juju.components.CopyToClipboard/>));
+    }
+    var c1, c2;
+    for (var i = 0, l = components.length; i < l; i++) {
+      c1 = ReactDOM.findDOMNode(components[i]).id;
+      assert.ok(c1, 'Component needs a unique ID');
+      for (var j = 0; j < l; j++) {
+        if (i === j) { continue; }
```

Comment:
If these are found out of order then this will be skipping two components which aren't actually the same one...